samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -441,15 +441,24 @@ public void afterStop() {
 
     @Override
     public void afterFailure(Throwable t) {
-      processors.removeIf(pair -> pair.getLeft().equals(processor));
-
+      // we need to close associated coordinator metadata store, although the processor failed
+      processors.forEach(sp -> {
+        if (sp.getLeft().equals(processor)) {
+          if (sp.getRight() != null) {
+            sp.getRight().close();
+          }
+          processors.remove(sp);
```

Comment:
Is it safe to modify (remove) the map here while iterating over it (foreach)?

Reply:
Thanks for pointing this. Changed to remove it after finishing iterating.